### PR TITLE
Refactored bluetooth enable/disable method RPCs

### PR DESCRIPTION
**Pull request by tprotopopov-dev**
This change should prevent bluetooth failures during turning On/Off.

---
This change is [Reviewable](https://reviewable.io/reviews/google/mobly-bundled-snippets/134)

### CODE REVIEW

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java [hunk 1]
```diff
@@ -333,4 +329,30 @@ public void onReceive(Context context, Intent intent) {
             }
         }
     }
+
+    /** Waits until bluetooth adapter state becomes consistent. */
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:333]: Add javadoc explaining the mechanism of this method.
Seems more complicated than a simple wait

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java [hunk 2]
```diff
@@ -205,24 +201,24 @@ public void btBecomeDiscoverable(Integer duration) throws Throwable {
         }
         // TODO(jwang1013): change it to SDK version for R after R is released.
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:202]: Can you fulfill this TODO now that R is released

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java [hunk 3]
```diff
@@ -89,47 +92,40 @@ private static BluetoothDevice getPairedDeviceByAddress(String deviceAddress) {
         return null;
     }
 
-    @Rpc(description = "Enable bluetooth with a 30s timeout.")
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:92]: why remove the timeout info?

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java [hunk 4]
```diff
@@ -333,4 +329,30 @@ public void onReceive(Context context, Intent intent) {
             }
         }
     }
+
+    /** Waits until bluetooth adapter state becomes consistent. */
+    private static void waitForStableBtState() throws BluetoothAdapterSnippetException {
+        Set<Integer> btStates = new HashSet<>();
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:335]: seems a bit too complicated to keep track of five ints. is it possible to simplify this logic?

**Review by xpconanfan — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 3 above)
Comment (tprotopopov-dev) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:92]: Timeout is irrelevant and can change.

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 1 above)
Comment (tprotopopov-dev) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java]: done

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 4 above)
Comment (tprotopopov-dev) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java]: Updated.

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 2 above)
Comment (tprotopopov-dev) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java]: Done

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 3 above)
Comment (xpconanfan) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:92]: It is hard coded and the users cannot change it, so we should not remove this?

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java [hunk 5]
```diff
@@ -333,4 +328,32 @@ public void onReceive(Context context, Intent intent) {
             }
         }
     }
+
+    /**
+     * Waits until the bluetooth adapter state becomes consistent. We consider BT state consistent
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:333]: "becomes consistent" -> "has stabilized"?

Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:333]: "state consistent" -> "state stabilized"

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java [hunk 6]
```diff
@@ -333,4 +328,32 @@ public void onReceive(Context context, Intent intent) {
...
+                        System.currentTimeMillis() + BT_MATCHING_STATE_INTERVAL_MS;
+            }
+            if (continuousStateIntervalMs <= System.currentTimeMillis()) {
+                return;
+            }
+            prevState = currentState;
+        }
+        throw new BluetoothAdapterSnippetException(
+                String.format(
+                        "Failed to reach stable Bluetooth state within %d ms",
```
Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:356]: "reach a stable"

Comment (xpconanfan) [inline src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java:356]: the unit should be "second" to be consistent with the other error messages.

**Review by xpconanfan — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 3 above)
Comment (tprotopopov-dev) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java]: Updated.

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 5 above)
Comment (tprotopopov-dev) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java]: Done.

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 6 above)
Comment (tprotopopov-dev) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java]: Done.

Comment (tprotopopov-dev) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java]: Done.

File: src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java (hunk 5 above)
Comment (tprotopopov-dev) [reply src/main/java/com/google/android/mobly/snippet/bundled/bluetooth/BluetoothAdapterSnippet.java]: Done.

**Review by xpconanfan — APPROVED** [review]

**Pull request merged**